doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+static inline ElektraKey * elektraKeyDup (ElektraKey * key, ElektraKeyCopyFlag flags)
+{
+    ElektraKey * dest = flags & ELEKTRA_KEY_CP_NAME
+        ? elektraKeyNew (elektraKeyGetName (key))
+        : elektraKeyNew (&(ElektraKeyname){.ns = ELEKTRA_NS_CASCADING, .name = "", .size = 0});
+    return elektraKeyCopy (dest, key, (ElektraKeyCopyFlag)(flags & ~ELEKTRA_KEY_CP_NAME));
+}
+
+
+ElektraKeyset * elektraKeysetNew (size_t prealloc);
+ElektraKeyset * elektraKeysetRetain (const ElektraKeyset * ks);
+void elektraKeysetRelease (const ElektraKeyset * ks);
+
+size_t elektraKeysetSize (ElektraKeyset * ks);
+
+ElektraErrorCode elektraKeysetInsert (ElektraKeyset * ks, ElektraKey * key);
+ElektraErrorCode elektraKeysetInsertAll (ElektraKeyset * ks, ElektraKeyset * other);
+static inline ElektraErrorCode elektraKeysetInsertAndRelease (ElektraKeyset * ks, ElektraKey * key)
+{
+    ElektraErrorCode error = elektraKeysetInsert (ks, key);
+    elektraKeyRelease (key);
+    return error;
+}
+
+ElektraKey * elektraKeysetGet (ElektraKeyset * ks, size_t index);
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
+
+void elektraKeysetRemove (ElektraKeyset * ks, size_t index);
+void elektraKeysetRemoveRange (ElektraKeyset * ks, size_t start, size_t end);
+static inline void elektraKeysetClear (ElektraKeyset * ks)
+{
+    elektraRemoveRange (ks, 0, elektraKeysetSize (ks));
+}
+
+size_t elektraKeysetLookup (ElektraKeyset * ks, const ElektraKeyname * name);
+
+size_t elektraKeysetFindHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
+```
+
+### Private headers
+
+```c
+struct ElektraKeynameCow {
+    struct ElektraKeyname name;
+
+    uint16_t refs;
+
+    bool shouldFree : 1;
+    int reserved : 15;
+};
+
+struct ElektraKeyvalueCow {
+    struct ElektraKeyvalue value;
+
+    uint16_t refs;
+
+    bool shouldFree : 1;
+    int reserved : 15;
+};
+
+struct ElektraKeyset;
+
+struct ElektraKey {
+    struct ElektraKeynameCow * name;
+    struct ElektraKeyvalueCow * value;
+    struct ElektraKeyset * meta;
+
+    uint16_t refs;
+    uint16_t nameLock;
+
+    bool shouldFree : 1;
+    bool needsSync : 1;
+    int reserved : 14;
+};
+
+struct ElektraKeyset {
+    struct ElektraKeysetDataCow * data;
+
+    uint16_t refs;
+
+    bool shouldFree : 1; // used for mmap
+    int reserved : 15;
+};
+```
+
+### Usage Examples
+
+```c
+// Create key (Note: better ways below via additional APIs)
+ElektraKey * k1 = elektraKeyNew (&(ElektraKeyname){
+    .ns = ELEKTRA_NS_SYSTEM,
+    .name = "foo\0bar\0baz",
+    .size = 14
+});
+
+// get name from key
+const ElektraKeyname * name1 = elektraKeyGetName (k1);
+
+// create local name
+ElektraKeyname name2 = {
+    .ns = ELEKTRA_NS_USER,
+    .name = "foo",
+    .size = 4
+};
+
+// ... using existing name
+ElektraKey * k2 = elektraKeyNew (&name2);
+
+// manipulate name
+elektraKeynamePushPart (&name2, "boo");
+ElektraKey * k3 = elektraKeyNew (&name2);
+
```

Comment:
example too long, can you please split up to small examples that show specific features of the API?

Reply:
That would just repeat the setup over and over again. I know the example is long, but that's why I added so many comments...

Reply:
You can assume the setup to be done, or (I would prefer) compilable C in separate files where you can have a function/macro for common setup.

In general I think it would be a good idea if the whole API and examples are compilable (maybe even implemented to some degree, e.g. for ref counting). This would lower the barrier to actually try out modifying your examples.

Reply:
Everything should compile already. Just put the API definition in a header (or into your test `.c` file) and try writing code. The _linker_ will of course complain because the functions aren't defined anywhere, but compiling should work...

More than that is not on the table right now. I won't invest the time to create an implementation that might not be used at all.